### ISSUE TITLE
AMP Elements ID enforcement will never work

### ISSUE THREAD

**Issue by mkhatib** [ampproject/amphtml#3938]
AMP automatically generate IDs and sets them to the element **before** build is called on the element. Hence logic that tries to enforce ID to be provided, like in `amp-user-notification` will never trigger because all AMP elements will have IDs by the time their `buildCallback` is called.

in `amp-user-notification.js`:

``` javascript
buildCallback() {
    ...
    this.elementId_ = user.assert(this.element.id,
        'amp-user-notification should have an id.');
   ...
}
```

in `resources-impl.js`

``` javascript
  add(element) {
    ...
    const resource = new Resource((++this.resourceIdCounter_), element, this);
    if (!element.id) {
      element.id = 'AMP_' + resource.getId();
    }
    ...
        resource.build();
    ...
}
```
## Potential Solutions
- Enforce publisher provided ID through the validator (should probably do this anyway for elements that this is a requirement for them)
- Reserve `AMP_` prefix in ID for AMP-generated IDs.
- Memoize element real ID before giving it an AMP-generated one.
- Move the logic for setting the generated ID after build is called (this is tricky and probably the most dangerous)

**Comment by cramforce** [comment]
Commented on other issue. As a quick fix, we should probably set a property on the DOM node when we generate an id.

**Comment by mkhatib** [comment]
I am worried this might break current usages, if we start enforcing publisher IDs now. Do we just add that and just report soft warnings for now?

**Comment by cramforce** [comment]
I think that is sufficiently unlikely. I'd definitely prefer if we changed the auto generated attributes to something like `amp-id` or just stop doing it.

**Comment by dvoytenko** [comment]
Not sure why we are setting an ID on AMP elements, besides possibly debug. A quick cheap fix might be to use `-amp-` prefix instead of current `AMP_` and when requiring IDs check that ID does not have `-amp-` prefix - or in other words it's a pub-generated ID. But I agree. We should see if we can stop doing that.

**Comment by jridgewell** [comment]
Closed by #6132.
